Gemfile

```diff
@@ -190,7 +190,7 @@ if ENV["ALLOW_DEV_POPULATE"] == "1"
   gem 'discourse_dev_assets'
   gem 'faker', "~> 2.16"
 else
-  group :development do
+  group :development, :test do
```

Comment:
Is there a reason we need this in the test environment now?

Reply:
I remember it was breaking specs, I believe it was related to loading files. I’m going to re-check that and see if this is still needed or not.

Reply:
Ok it’s breaking the annotations job on our CI:
```
/__w/discourse/discourse/vendor/bundle/ruby/2.7.0/gems/bootsnap-1.11.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:15:in `require': cannot load such file -- faker (LoadError)
	from /__w/discourse/discourse/vendor/bundle/ruby/2.7.0/gems/bootsnap-1.11.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:15:in `require'
	from /__w/discourse/discourse/vendor/bundle/ruby/2.7.0/gems/zeitwerk-2.5.4/lib/zeitwerk/kernel.rb:35:in `require'
	from /__w/discourse/discourse/lib/faker/discourse.rb:3:in `<main>'
	from /__w/discourse/discourse/vendor/bundle/ruby/2.7.0/gems/bootsnap-1.11.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	from /__w/discourse/discourse/vendor/bundle/ruby/2.7.0/gems/bootsnap-1.11.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	from /__w/discourse/discourse/vendor/bundle/ruby/2.7.0/gems/zeitwerk-2.5.4/lib/zeitwerk/kernel.rb:27:in `require'
	from /__w/discourse/discourse/vendor/bundle/ruby/2.7.0/gems/zeitwerk-2.5.4/lib/zeitwerk/loader/helpers.rb:95:in `const_get'
	from /__w/discourse/discourse/vendor/bundle/ruby/2.7.0/gems/zeitwerk-2.5.4/lib/zeitwerk/loader/helpers.rb:95:in `cget'
	from /__w/discourse/discourse/vendor/bundle/ruby/2.7.0/gems/zeitwerk-2.5.4/lib/zeitwerk/loader.rb:237:in `block (2 levels) in eager_load'
	from /__w/discourse/discourse/vendor/bundle/ruby/2.7.0/gems/zeitwerk-2.5.4/lib/zeitwerk/loader/helpers.rb:26:in `block in ls'
	from /__w/discourse/discourse/vendor/bundle/ruby/2.7.0/gems/zeitwerk-2.5.4/lib/zeitwerk/loader/helpers.rb:18:in `each_child'
	from /__w/discourse/discourse/vendor/bundle/ruby/2.7.0/gems/zeitwerk-2.5.4/lib/zeitwerk/loader/helpers.rb:18:in `ls'
	from /__w/discourse/discourse/vendor/bundle/ruby/2.7.0/gems/zeitwerk-2.5.4/lib/zeitwerk/loader.rb:232:in `block in eager_load'
	from /__w/discourse/discourse/vendor/bundle/ruby/2.7.0/gems/zeitwerk-2.5.4/lib/zeitwerk/loader.rb:217:in `synchronize'
	from /__w/discourse/discourse/vendor/bundle/ruby/2.7.0/gems/zeitwerk-2.5.4/lib/zeitwerk/loader.rb:217:in `eager_load'
	from /__w/discourse/discourse/vendor/bundle/ruby/2.7.0/gems/railties-7.0.2.3/lib/rails/autoloaders.rb:32:in `each'
	from /__w/discourse/discourse/vendor/bundle/ruby/2.7.0/gems/railties-7.0.2.3/lib/rails/application.rb:497:in `eager_load!'
	from /__w/discourse/discourse/vendor/bundle/ruby/2.7.0/gems/railties-7.0.2.3/lib/rails/engine.rb:358:in `eager_load!'
	from /__w/discourse/discourse/vendor/bundle/ruby/2.7.0/gems/annotate-3.2.0/lib/annotate.rb:98:in `eager_load'
	from /__w/discourse/discourse/vendor/bundle/ruby/2.7.0/gems/annotate-3.2.0/bin/annotate:29:in `<top (required)>'
	from bin/annotate:29:in `load'
	from bin/annotate:29:in `<main>'
```

Reply:
Hmm can we dig a little deeper into this? The annotate job does not have a dependency on either `discourse_dev_assets` or `faker`. 

```
    annotate (3.2.0)
      activerecord (>= 3.2, < 8.0)
      rake (>= 10.4, < 14.0)
```

So we got to figure out why faker is suddenly being loaded in the test env now.

Reply:
O it looks like `lib/faker/discourse` is being autoloaded causing this to be loaded in the test env.

Reply:
Actually `annotate` will eager load the whole app when annotations for models are enabled: https://github.com/ctran/annotate_models/blob/develop/bin/annotate#L29

So that’s why. If we don’t want this, then I guess we can just check if `Faker` is defined in `lib/faker/discourse` and skip the whole file if it isn’t.

Reply:
OK I see what the problem is now. Basically, `lib/faker/discourse` should not be part of the autoload path and should be manually required when we need it for the rake task.

Reply:
If I remember correctly, I don't think all files in the `lib` directory and its sub-directories are autoloaded by default. Perhaps we can make `lib` not autoloaded but `app/lib` autoloaded? Kind of stealing ideas from this tweet I saw: 

```
/lib - utilities like; tasks, generators, load test scripts, migration backfills, development seeds and such.

/app/lib - production code like; 3rd party integrations and any code that could be gemified or used in other apps.
```

Reply:
It’s even more than that, by default Rails doesn’t autoload `lib` at all. Usually the code that lies in here must be required explicitly.
If that’s not a problem of moving everything around like you’re suggesting, then I think we should try to do it somehow. Leave stuff we know should not be autoloaded at all in `lib` and move everything else in `app/lib` for the time being. `app/lib` will be autoloaded automatically by Rails this way.
It’s again more work and it’s going to be a bit painful but I guess we could benefit a lot from this :grin:

Reply:
Yea but this is going to be a decision that is hard to revert. Can you perhaps spin up a topic on dev so we give others a chance to give their input?